src/events/balancer.py

```diff
@@ -199,8 +199,10 @@ def rebalance(self, event: ActionEvent) -> None:
         if self.charm.state.runs_broker:
             available_brokers = [broker.unit_id for broker in self.charm.state.brokers]
         else:
-            brokers: list = self.charm.state.balancer.broker_capacities.get("brokerCapacities", [])
-            available_brokers = [int(broker["brokerId"]) for broker in brokers]
+            brokers: list = [
+                broker.name for broker in getattr(self.charm.state.balancer.relation, "units", [])
+            ]
```

Comment:
**nitpick:** Although we do make liberal use of `getattr`, I think it might be better to do the following to preserve typing:

```python
brokers = [broker.name for broker in self.charm.state.balancer.relation.units] if self.charm.state.balancer.relation else []

Reply:
Agreed, it's better. Changed in a6e2efbc3484